File: VirtualPetC#/VirtualPetC#/Core/Pets/Dog.cs
namespace VirtualPetC_.Core.Pets;

/// <summary>
/// Dog class 
/// Inherits from abstract Pet class and provides unique dog-specific behaviors.
/// Special Ability: Loyalty - maintains happiness longer than other pets.
/// </summary>
public class Dog : BasePet
{
    private bool loyaltyActive;
    private DateTime loyaltyEndTime;

    // Constructor - calls base Pet constructor
    public Dog(string name) : base(name)
    {
        loyaltyActive = false;
        loyaltyEndTime = DateTime.MinValue;
    }

    public bool IsLoyaltyActive => loyaltyActive;

    
    // Implementation of MakeSound Woof Woof
    public override void MakeSound()
    {
        Console.WriteLine($"{Name} says: Woof! Woof!");
        Happiness += 5;
        Console.WriteLine($"{Name} wags tail happily!");
    }

    
    
    // Returns description of dog's special ability.
    public override string GetSpecialAbility()
    {
        string status = loyaltyActive ? "(Active)" : "(Ready)";
        return $"Loyalty - Reduces happiness decay! {status}";
    }

    
    
    // Dogs get extra Happiness from Play
        public override string Play()
    {
        // Dogs get MORE happiness from playing due to their playful nature
        Happiness += 25;  // Base Pet gives 20, dogs get 25
        Hunger -= 10;  // Playing makes pet hungry

        return $"{Name} is playing fetch! \n{Name} brings the ball back with tail wagging!\n{Name} is very happy but getting hungry from all that running!";
    }

    
    // Override to reduce happiness decay when loyalty is active.
    protected override double GetHappinessDecayModifier()
    {
        return loyaltyActive ? LoyaltyHappinessReduction : 1.0;
    }

    
    // Special dog ability: Activate loyalty to reduce happiness decay.
    public string UseSpecialAbility()
    {
        loyaltyActive = true;
        loyaltyEndTime = DateTime.Now.AddSeconds(LoyaltyDuration);

        return $"{Name} is feeling extra loyal! Happiness will decay slower for the next 60 seconds.";
    }

    
    // Check if the dog can use its special ability.
    public bool CanUseAbility()
    {
        return !loyaltyActive; // Can only use when loyalty is not currently active
    }

    
    /// Update method to check if loyalty has expired.
    public override void Update(double deltaTime)
    {
        // Call base update (which uses our modifier)
        base.Update(deltaTime);

        // Check if loyalty has expired
        if (loyaltyActive && DateTime.Now >= loyaltyEndTime)
        {
            loyaltyActive = false;
        }
    }
}

File: VirtualPetGo/pet/dog.go
package pet

import "time"

type Dog struct {
	BasePet
	loyaltyActive  bool
	loyaltyEndTime time.Time
}

func NewDog(name string) *Dog {
	return &Dog{
		BasePet:       newBasePet(name),
		loyaltyActive: false,
	}
}

func (d *Dog) Play() string {
	// Call base play behaviour
	d.BasePet.Play()

	//Add extra happiness for fetch
	d.setHappiness(d.GetHappiness() + 20)
	d.setHunger(d.GetHunger() - 10)

	return d.GetName() + " loves playing fetch! Extra happiness gained"
}

func (d *Dog) MakeSound() string {
	return "Woof! Woof!"
}
func (d *Dog) Interact() string {
	return d.GetName() + " says " + d.MakeSound()
}
func (d *Dog) getHappinessDecayModifier() float64 {
	if d.loyaltyActive {
		return LoyaltyHappinessReduction // 0.5
	}
	return 1.0
}

func (d *Dog) Update(deltaTime float64) {
	// Call base update (which uses our modifier)
	d.BasePet.Update(deltaTime)

	// Check if loyalty has expired
	if d.loyaltyActive && time.Now().After(d.loyaltyEndTime) {
		d.loyaltyActive = false
	}
}
func (d *Dog) UseSpecialAbility() string {
	d.loyaltyActive = true
	d.loyaltyEndTime = time.Now().Add(time.Duration(LoyaltyDuration) * time.Second)

	return d.GetName() + " is feeling extra loyal! Happiness will decay slower for the next 60 seconds."
}
func (d *Dog) CanUseAbility() bool {
	return !d.loyaltyActive // Can only use when loyalty is not currently active
}
func (d *Dog) GetStatus() Status {
	// Determine ability status text
	abilityStatus := "(Inactive)"
	if d.loyaltyActive {
		abilityStatus = "(Active)"
	}

	return Status{
		Name:     d.GetName(),
		Type:     "Dog",
		Age:      d.GetAge(),
		AgeStage: d.BasePet.getAgeStage(),

		// Core stats
		Health:      d.GetHealth(),
		Hunger:      d.GetHunger(),
		Happiness:   d.GetHappiness(),
		Cleanliness: d.GetCleanliness(),

		// Special ability info
		SpecialAbility: "Loyalty - Maintains happiness longer!",
		AbilityStatus:  abilityStatus,

		// Overall status
		StatusMessage: d.BasePet.getStatusMessage(),
		IsAlive:       d.IsAlive(),
		// Illness status
		IsIll:       d.IsIll(),
		IllnessName: d.GetIllness(),
	}
}
